keps/sig-storage/20190204-scale-testing.md

```diff
@@ -107,7 +146,7 @@ These tests should measure how long it takes to start up a pod with volumes, ass
 
 For each volume type:
 
-* Create many pods with 1 volume each.
+* Create many pods with 1 volume each in parallel.
```

Comment:
Some more points to consider:
* Do we create one volume per pod? Is that realistic (i.e. can the storage system scale to that number)? If we don't and volumes cannot be shared arbitrarily, then pods become unschedulable.
* The proposed test scales all pods evenly across all nodes, right? As a side effect, we get volumes also spread evenly. I am wondering whether it makes sense to have a test that stresses scalability of the local operations by running a high number of pods+volumes on a single node.

Reply:
> * Do we create one volume per pod? Is that realistic (i.e. can the storage system scale to that number)? If we don't and volumes cannot be shared arbitrarily, then pods become unschedulable.

If you look into our current tests for stateless, what we do is:
- saturate the cluster as fast as we can (when we don't really measure anything)
- then generate a constant stream of pods (actually we don't create them in parallel) for which we measure stuff
I think that would address your concern?

> * The proposed test scales all pods evenly across all nodes, right? As a side effect, we get volumes also spread evenly. I am wondering whether it makes sense to have a test that stresses scalability of the local operations by running a high number of pods+volumes on a single node.

+1
Though it's a slightly different kind of test, that is enough to have a 1-node cluster to run.

Reply:
What does "saturate the cluster" mean? Can you point me to the code which does it or a description?

The actual measurements are then done sequentially (create one pod, wait for it to be running, stop it, create next pod, etc.)?

That would work much better for pods using the same PV. But timing will vary greatly depending on the scheduling pattern: one pod running after another on the same node will start up much faster than a pod for which the volume first has to be moved from the old node to the new node.

Reply:
I don't want to handle sharing PVs in the first phase at least, because how it behaves is highly variable on the storage backend type (RWO vs RWX) and how it gets scheduled.

What does saturating the cluster achieve? Is it for pre-pulling images onto the nodes?

I will add more detail that for these initial tests, there is no sharing of volumes, and we can stress various dimensions:

* Total number of volumes per pod (1 pod with many volumes on one node)
* Total number of volumes per node (many pods with 1 volume each on one node)
* Total number of volumes per cluster (can we take total # volumes per node and multiply by # nodes?)